include/boost/math/special_functions/detail/interval_prime_sieve.hpp

```diff
@@ -0,0 +1,337 @@
+            178'936'222'537'081, 696'161'110'209'049, 2'854'909'648'103'881, 6'450'045'516'630'769, 11'641'399'247'947'921,
+            190'621'428'905'186'449, 196'640'148'121'928'601, 712'624'335'095'093'521, 1'773'855'791'877'850'321,
+            2'327'687'064'124'474'441, 6'384'991'873'059'836'689, 8'019'204'661'305'419'761, 10'198'100'582'046'287'689u,
+
+            (static_cast<int_128t>(0x3uLL) << 64) | 0xc956f827e0524359uLL,      // 69'848'288'320'900'186'969
+            (static_cast<int_128t>(0xbuLL) << 64) | 0x539315b3b1268d59uLL,      // 208'936'365'799'044'975'961
+            (static_cast<int_128t>(0x1cuLL) << 64) | 0xec87d86ca60b50a1uLL,     // 533'552'663'339'828'203'681
+            (static_cast<int_128t>(0x32uLL) << 64) | 0xc6d3496f20db3d81uLL,     // 936'664'079'266'714'697'089
+            (static_cast<int_128t>(0x74uLL) << 64) | 0x210967a12ba94be1uLL,     // 2'142'202'860'370'269'916'129
+            (static_cast<int_128t>(0x2e3uLL) << 64) | 0xec11ddc09fd65c51uLL,    // 13'649'154'491'558'298'803'281
+            (static_cast<int_128t>(0x753uLL) << 64) | 0x641c14b397c27bf1uLL,    // 34'594'858'801'670'127'778'801
+            (static_cast<int_128t>(0x1511uLL) << 64) | 0x85fdf38d1fc9ce21uLL,   // 99'492'945'930'479'213'334'049
+            (static_cast<int_128t>(0x3e8buLL) << 64) | 0xaba417e222ca5091uLL    // 295'363'187'400'900'310'880'401
+        };
+    };
+
+    static constexpr pssentry pss_{};
+    static constexpr boost::math::detail::prime_sieve::MOD210Wheel<Integer> w_{};
+    std::size_t tdlimit_;
+
+    Integer delta_;
+    Integer left_;
+    Integer right_;
+
+    OutputIterator resultant_primes_;
+
+    // https://www.researchgate.net/publication/220803585_Performance_of_C_bit-vector_implementations
+    boost::dynamic_bitset<> b_;
+    
+    std::vector<Integer> primes_;
+    std::int_fast64_t plimit_;
+
+    void Settdlimit() noexcept;
+    void SeiveLength(const Integer d) noexcept;
+    void Sieve() noexcept;
+    bool Psstest(const std::size_t pos) noexcept;
+    void Psstestall() noexcept;
+    decltype(auto) WriteOutput() noexcept;
+    
+public:
+    IntervalSieve(const Integer left, const Integer right, OutputIterator resultant_primes) noexcept;
+    decltype(auto) NewRange(const Integer left, const Integer right) noexcept;
+    decltype(auto) NewRange(const Integer left, const Integer right, OutputIterator resultant_primes) noexcept;
+};
+
+template<typename Integer, typename OutputIterator>
+void IntervalSieve<Integer, OutputIterator>::Settdlimit() noexcept
+{
+    const double dr {static_cast<double>(right_)};
+    const double delta {static_cast<double>(delta_)};
+    const double tdest {delta * std::log(dr)};
+
+    // Small cases
+    if(tdest * tdest >= dr)
+    {
+        tdlimit_ = static_cast<std::size_t>(std::sqrt(dr));
+        plimit_ = 0;
+        return;
+    }
+
+    // First guess
+    if(tdest <= 1ul<<30)
+    {
+        tdlimit_ = static_cast<std::size_t>(tdest);
```

Comment:
As a complete aside, but are all these `static_cast`s really necessary? I know there are some tricky rules about conversions sometimes, but I just didn't imagine it was necessary to be _explicit_ so often. If the implicit cast is correct, it makes for much easier (less noisy) code to read.

Reply:
Unfortunately they are. I do not claim to know what is going on under the hood of `multiprecision::gmp_int` but it does not seem to implicitly cast. Almost every explicit cast is to fix a compiler error for `gmp_int`.

Reply:
They are code clutter, but serve to document conversions that probably lose precision, saying "OK - I don't care."

Implicit conversion is a big pit with spikes at the bottom waiting to catch the unwary and is a major cause of very-difficult-to-trace loss of precision using Boost.Multiprecision.

(And I have fallen into it too many times... ouch!)

Reply:
Aha, yeah, OK, thanks.